src/v/cluster/members_backend.cc

```diff
@@ -593,10 +598,30 @@ void members_backend::calculate_reallocations_batch(
                 continue;
             }
         }
+        size_t rand_idx = 0;
+        /**
+         * We use 64 bit field initialized with random number to sample topic
+         * partitions to move, every time the bitfield is exhausted we
+         * reinitialize it and reset counter. This way we chose random
+         * partitions to move without iterating multiple times over the topic
+         * set
+         */
+        std::bitset<sizeof(uint64_t)> sampling_bytes;
         for (const auto& p : metadata.get_assignments()) {
             if (idx_it->second > current_idx++) {
                 continue;
             }
+
+            if (rand_idx % sizeof(uint64_t) == 0) {
+                sampling_bytes = random_generators::get_int(
+                  std::numeric_limits<uint64_t>::max());
+                rand_idx = 0;
+            }
+
+            if (sampling_bytes.test(rand_idx++)) {
```

Comment:
Wouldn't this method lead to a bernoulli distribution where partitions that are iterated through earlier are more likely to be in the final result set than later partitions?

I.e, we're generating a sequence of partitions with this method. If the whole sequence was the result then each partition would be uniformly likely to be in that sequence. However, we really only take the first N partitions from the sequence which leads to the non-uniform distribution.

I think the goal is to have the probability that each partition moves to the new node be uniform. To that ends it may be easier to randomly select (topic, partition) indices until we have N partitions to move. Though this would require generating a lot more random numbers.

Reply:
> I.e, we're generating a sequence of partitions with this method. If the whole sequence was the result then each partition would be uniformly likely to be in that sequence. However, we really only take the first N partitions from the sequence which leads to the non-uniform distribution.

this is only partially true, we continue where we left of if we reached max move batch size we skip over to the partition we left off using `last_ntp_index`

Reply:
It looks like we're still selecting N replicas from each existing node in the cluster to move to the newly added node. Maybe things are close enough to a uniform distribution not to matter much though.

Reply:
i did some testing and indeed it looks like it is close to the uniform distribution with current approach.